File: src/collatz.exs
collatz = fn num, func ->
    IO.puts to_string num
    if rem(num, 2) == 1 do
        unless num == 1 do
            func.(3 * num + 1, func)
        else 1 end
    else
        func.(div(num, 2), func)
    end
end

collatz.(9, collatz)

File: src/collatz.java
class Collatz
{
    public static int Collatz(int n)
    {
        System.out.println(n);
        if (n == 1)
            return 1;
        if (n % 2 == 1)
            return Collatz(3 * n + 1);
        return Collatz(n / 2);
    }

    public static void main(String[] args)
    {
        System.out.println(Collatz(9));
    }
}
